In [3]:
import pandas as pd

# importando os dois datasets para a classe DataFrame do Pandas

df_quiz = pd.read_csv('./case estágio de dados/quiz-sample.csv', parse_dates=['data_submissao'])
df_orders = pd.read_csv('./case estágio de dados/orders-sample.csv', parse_dates=['processed_at', 'cancelled_at'])

df_quiz

person_id  genero  \
0       z85vnNzPszkToYzZaCiiqguiaoRFey3ssxUyrla0KwU=  Mulher   
1       qJX1qjRTIrxsLRk4Altj4fG2/UQi2A0DFyfUk7jnL6c=   Homem   
2       15amHiG6TcaEqYX7irM7vWPhkrJWcQ+DEUzKPMkGLVc=   Homem   
3       MRzwwVtTWEJk7c93frTnXuSY7/VHrMZQBQxslvoAKiw=  Mulher   
4       MMYx9aY/oQljcIN1bL7e2/z1whUpHPPTX+8TBJPAWzM=  Mulher   
...                                              ...     ...   
213019  nGzZyKQrOMWr0iQ4gE//Gx8UK+10VTadD9R+WHIjraQ=  Mulher   
213020  O9hfKxZ1IeEz4sgr/nncEW1TeIRQtOYbx3oUTPVnWY4=  Mulher   
213021  2T4dtiohhaXXOucpsBEA+VhiUG2RTSEeup9hYAuoeMo=  Mulher   
213022  klspdapSdyvVmOLBNLQH/ob4K7G1UD12xbMKy9CKXBg=  Mulher   
213023  RAjxQiRQNjiAJe9P0xtUU/B2oil4sgYCFDKicJCGqS0=  Mulher   

                     idade tipo_pele  pele_sensivel             estado_vive  \
0       Entre 18 e 24 anos     Mista              0  Rio Grande do Sul - RS   
1       Entre 18 e 24 anos    Normal              0            Paraíba - PB   
2       Entre 18 e 24 anos    Normal              1              Bahia - BA   
3       Entre 18 e 24 anos     Mista              1             Paraná - PR   
4       Entre 18 e 24 anos   Não sei              1             Paraná - PR   
...                    ...       ...            ...                     ...   
213019  Entre 35 e 39 anos    Oleosa              0          São Paulo - SP   
213020  Entre 35 e 39 anos    Oleosa              0               Pará - PA   
213021  Entre 45 e 49 anos    Oleosa              0     Espírito Santo - ES   
213022  Entre 40 e 44 anos     Mista              1     Espírito Santo - ES   
213023  Entre 35 e 39 anos     Mista              1          São Paulo - SP   

       data_submissao        preocupacao_de_pele  tem_preocupacao  
0          2021-07-11     pp_bolsas_abaixo_olhos                0  
1          2021-07-06     pp_bolsas_abaixo_olhos                1  
2          2021-06-30     pp_bolsas_abaixo_olhos                0  
3          2021-06-25     pp_bolsas_abaixo_olhos                1  
4          2021-06-24     pp_bolsas_abaixo_olhos                0  
...               ...                        ...              ...  
213019     2019-04-11  pp_nao_tenho_preocupacoes                0  
213020     2020-04-09  pp_nao_tenho_preocupacoes                0  
213021     2020-02-24  pp_nao_tenho_preocupacoes                0  
213022     2019-09-17  pp_nao_tenho_preocupacoes                0  
213023     2019-09-02  pp_nao_tenho_preocupacoes                0  

[213024 rows x 9 columns]

In [4]:
df_orders

product_A  product_B  product_C  product_D  product_E  \
0              0          0          1          0          0   
1              0          0          1          1          0   
2              0          0          0          1          0   
3              0          0          2          0          0   
4              0          1          1          0          0   
...          ...        ...        ...        ...        ...   
30226          0          0          2          0          0   
30227          1          0          0          0          0   
30228          1          0          1          1          0   
30229          1          1          1          1          0   
30230          0          2          0          0          0   

                   processed_at              cancelled_at financial_status  \
0     2020-01-01 08:02:29+00:00                       NaT             paid   
1     2020-01-02 15:14:49+00:00                       NaT             paid   
2     2020-05-18 08:16:30+00:00                       NaT             paid   
3     2020-05-29 12:01:25+00:00                       NaT             paid   
4     2020-05-21 14:51:49+00:00                       NaT             paid   
...                         ...                       ...              ...   
30226 2020-08-12 20:20:48+00:00                       NaT             paid   
30227 2020-08-10 17:14:57+00:00 2020-08-18 10:04:25+00:00          pending   
30228 2020-08-10 13:49:39+00:00 2020-08-18 10:03:43+00:00          pending   
30229 2020-08-13 18:54:57+00:00                       NaT             paid   
30230 2020-05-18 20:10:15+00:00 2020-05-26 10:04:30+00:00          pending   

                                          person_id          billing_city  \
0      QmxkQi1OVAJCnHhfd7lRP7uRkniwxjo51BR6i4lPk/4=             São Paulo   
1      aOSgAYOuG05sldfAHhP9EJAPoaRXHz3MApCvz/ywuMY=             São Paulo   
2      UNcuZslJ0yQVIH7ZHqEw8imWvmGOHVEwWmo1m2Y11yE=  Aparecida do Taboado   
3      a/HhG1GeKoJ4/HWNSYg0aIJjwGvQuacN8xzn1ETbNh8=               Jundiaí   
4      JoyQclINSH2gCcTNjONezpqC1T1lnx0fbB9f2eDvTVc=             Paranaguá   
...                                             ...                   ...   
30226  nabYxzWdQ9hwAeA/BGLyBM36Z14686k8zWoKWaeRSWI=             Cabo Frio   
30227  XAJ/ltJjd+HKIR4O4GQtaYlUxgmfeyMnHxEVSsxVLWY=             São Paulo   
30228  V8zUkC/Gvbigh6qrdgneIOdHbT8GnB343675qz2n08E=       Taboão da Serra   
30229  +rEtOYX7cGiGS5fGCQstN2pRVCWp4s6KCXFy6+qRh5g=              Brasília   
30230  joLvsbALaaCqC/WR+eNx+CvWKD9mQLbIhShaqmZ/YB0=                Canoas   

            billing_state  
0               São Paulo  
1               São Paulo  
2      Mato Grosso do Sul  
3               São Paulo  
4                  Paraná  
...                   ...  
30226      Rio de Janeiro  
30227           São Paulo  
30228           São Paulo  
30229    Distrito Federal  
30230   Rio Grande do Sul  

[30231 rows x 11 columns]

In [5]:
df_quiz.groupby([df_quiz['data_submissao'].dt.year, df_quiz['data_submissao'].dt.month, 'preocupacao_de_pele'])[['preocupacao_de_pele']].count().rename(columns={'preocupacao_de_pele': 'Total'})

Total
data_submissao data_submissao preocupacao_de_pele                    
2019           3              pp_bolsas_abaixo_olhos              145
                              pp_cicatrizes_espinha               145
                              pp_cicatrizes_machucado_cirurgia    145
                              pp_cravos                           145
                              pp_descamacao                       145
...                                                               ...
2021           7              pp_rosacea                           62
                              pp_rugas                             62
                              pp_sardas                            62
                              pp_vasinhos_aparentes                62
                              pp_vermelhidao                       62

[696 rows x 1 columns]

In [6]:
# ver quantas preocupações existiam no DataFrame original

preocupacoes = df_quiz[['preocupacao_de_pele']].drop_duplicates().sort_values('preocupacao_de_pele').set_index('preocupacao_de_pele')

preocupacoes

Empty DataFrame
Columns: []
Index: [pp_bolsas_abaixo_olhos, pp_cicatrizes_espinha, pp_cicatrizes_machucado_cirurgia, pp_cravos, pp_descamacao, pp_excesso_brilho, pp_flacidez_rosto, pp_linhas_finas, pp_manchas_brancas, pp_manchas_escuras, pp_melasma, pp_muitas_espinhas, pp_nao_tenho_preocupacoes, pp_oleosidade, pp_olheiras, pp_pintas, pp_poros_dilatados, pp_poucas_espinhas, pp_ressecamento, pp_rosacea, pp_rugas, pp_sardas, pp_vasinhos_aparentes, pp_vermelhidao]

In [7]:
# criando um novo DataFrame com os dados da person_id relacionados de uma tabela na outra

df_merge = pd.merge(df_quiz, df_orders, on=['person_id'], how='inner') # .drop_duplicates(['person_id', 'preocupacao_de_pele'])

# merge -> faz a unificação dos dados que possuem relação, neste caso, person_id dos dois DataFrames
# drop_duplicates -> para remover pessoas duplicadas do DataFrame (existiam pessoas que tinha uma mesma preocupação, então retirei
#                    as que possuiam essas duplicações)
# reset_index -> para colocar os index em ordem

df_merge

person_id  genero  \
0       6bobH7lLMcVLVmXd5RYZWQCNtw/NNfb6CG1nvT/1pfc=  Mulher   
1       6bobH7lLMcVLVmXd5RYZWQCNtw/NNfb6CG1nvT/1pfc=  Mulher   
2       6bobH7lLMcVLVmXd5RYZWQCNtw/NNfb6CG1nvT/1pfc=  Mulher   
3       6bobH7lLMcVLVmXd5RYZWQCNtw/NNfb6CG1nvT/1pfc=  Mulher   
4       6bobH7lLMcVLVmXd5RYZWQCNtw/NNfb6CG1nvT/1pfc=  Mulher   
...                                              ...     ...   
210955  RAjxQiRQNjiAJe9P0xtUU/B2oil4sgYCFDKicJCGqS0=  Mulher   
210956  RAjxQiRQNjiAJe9P0xtUU/B2oil4sgYCFDKicJCGqS0=  Mulher   
210957  RAjxQiRQNjiAJe9P0xtUU/B2oil4sgYCFDKicJCGqS0=  Mulher   
210958  RAjxQiRQNjiAJe9P0xtUU/B2oil4sgYCFDKicJCGqS0=  Mulher   
210959  RAjxQiRQNjiAJe9P0xtUU/B2oil4sgYCFDKicJCGqS0=  Mulher   

                     idade tipo_pele  pele_sensivel            estado_vive  \
0       Entre 25 e 29 anos   Não sei              0  Distrito Federal - DF   
1       Entre 25 e 29 anos   Não sei              0  Distrito Federal - DF   
2       Entre 25 e 29 anos   Não sei              0  Distrito Federal - DF   
3       Entre 25 e 29 anos   Não sei              0  Distrito Federal - DF   
4       Entre 25 e 29 anos   Não sei              0  Distrito Federal - DF   
...                    ...       ...            ...                    ...   
210955  Entre 35 e 39 anos     Mista              1         São Paulo - SP   
210956  Entre 35 e 39 anos     Mista              1         São Paulo - SP   
210957  Entre 35 e 39 anos     Mista              1         São Paulo - SP   
210958  Entre 35 e 39 anos     Mista              1         São Paulo - SP   
210959  Entre 35 e 39 anos     Mista              1         São Paulo - SP   

       data_submissao               preocupacao_de_pele  tem_preocupacao  \
0          2021-06-22            pp_bolsas_abaixo_olhos                0   
1          2021-06-22             pp_cicatrizes_espinha                0   
2          2021-06-22  pp_cicatrizes_machucado_cirurgia                0   
3          2021-06-22                         pp_cravos                1   
4          2021-06-22                     pp_descamacao                0   
...               ...                               ...              ...   
210955     2019-09-02             pp_vasinhos_aparentes                0   
210956     2019-09-02                    pp_vermelhidao                0   
210957     2019-09-02                    pp_vermelhidao                0   
210958     2019-09-02         pp_nao_tenho_preocupacoes                0   
210959     2019-09-02         pp_nao_tenho_preocupacoes                0   

        product_A  product_B  product_C  product_D  product_E  \
0               0          0          1          0          0   
1               0          0          1          0          0   
2               0          0          1          0          0   
3               0          0          1          0          0   
4               0          0          1          0          0   
...           ...        ...        ...        ...        ...   
210955          0          0          1          1          0   
210956          0          1          0          1          0   
210957          0          0          1          1          0   
210958          0          1          0          1          0   
210959          0          0          1          1          0   

                    processed_at cancelled_at financial_status  \
0      2020-04-26 11:10:32+00:00          NaT             paid   
1      2020-04-26 11:10:32+00:00          NaT             paid   
2      2020-04-26 11:10:32+00:00          NaT             paid   
3      2020-04-26 11:10:32+00:00          NaT             paid   
4      2020-04-26 11:10:32+00:00          NaT             paid   
...                          ...          ...              ...   
210955 2020-03-02 15:34:11+00:00          NaT             paid   
210956 2020-04-29 20:46:15+00:00          NaT             paid   
210957 2020-03-02 15:34:11+00:00          NaT    

In [8]:
df_merge.groupby(df_merge['data_submissao'].dt.year).size().reset_index().set_index('data_submissao')

0
data_submissao        
2019             87384
2020            104400
2021             19176

In [9]:
df_merge.groupby([df_merge['data_submissao'].dat.year, 'preocupacao_de_pele'])[['preocupacao_de_pele']].count().rename(columns={'preocupacao_de_pele': 'Total'})

Total
data_submissao preocupacao_de_pele                    
2019           pp_bolsas_abaixo_olhos             3641
               pp_cicatrizes_espinha              3641
               pp_cicatrizes_machucado_cirurgia   3641
               pp_cravos                          3641
               pp_descamacao                      3641
...                                                ...
2021           pp_rosacea                          799
               pp_rugas                            799
               pp_sardas                           799
               pp_vasinhos_aparentes               799
               pp_vermelhidao                      799

[72 rows x 1 columns]

In [10]:
# agrupa por ano e mês e conta quantas preocupações existem em cada mês

df_quant = df_merge.groupby([df_merge['data_submissao'].dt.year, df_merge['data_submissao'].dt.month, 'preocupacao_de_pele'])[['preocupacao_de_pele']].count().rename(columns={'preocupacao_de_pele': 'Total'})

df_quant

Total
data_submissao data_submissao preocupacao_de_pele                    
2019           3              pp_bolsas_abaixo_olhos              219
                              pp_cicatrizes_espinha               219
                              pp_cicatrizes_machucado_cirurgia    219
                              pp_cravos                           219
                              pp_descamacao                       219
...                                                               ...
2021           7              pp_rosacea                           47
                              pp_rugas                             47
                              pp_sardas                            47
                              pp_vasinhos_aparentes                47
                              pp_vermelhidao                       47

[696 rows x 1 columns]

In [51]:
# exporta para um arquivo para visualizar inteiro

df_quant.to_string('output.txt')

## B) Considerando as bases de quiz e compras, responda:

### 1. Quais as 5 preocupações de pele mais recorrentes entre os nossos clientes a cada mês, considerando apenas respostas de pessoas que possuam pedidos registrados na base de compras?

Observando os dados e fazendo uma análise em cada mês de quantas preocupações existiram, cheguei ao resultado de que todas elas equiparam-se, acredito que tenho feito de forma correta a análise, desta forma todas as preocupações obtiveram valores iguais em todos os meses.

In [11]:
df_mark = pd.read_csv('./case estágio de dados/product-marketing-investment.csv')

df_mark

report_month    product  investment
0       2020-01  product_C  28693.9200
1       2020-01  product_D   6186.4400
2       2020-02  product_C  45922.0700
3       2020-02  product_D   6389.7800
4       2020-03  product_C  32820.3900
5       2020-03  product_D  15922.6700
6       2020-04  product_B   8596.7000
7       2020-04  product_C  29192.6000
8       2020-04  product_D   9870.9600
9       2020-05  product_B  31033.6600
10      2020-05  product_C  19019.0000
11      2020-05  product_D   4301.9500
12      2020-06  product_A  53457.5200
13      2020-06  product_B  14666.4500
14      2020-06  product_C  20794.9700
15      2020-06  product_D   4055.8400
16      2020-07  product_A  70477.4000
17      2020-07  product_B  20615.9800
18      2020-07  product_C  48491.8800
19      2020-07  product_D  21803.1800
20      2020-07  product_E      0.0000
21      2020-08  product_A  48659.8300
22      2020-08  product_B  17372.8500
23      2020-08  product_C  40238.8025
24      2020-08  product_D  15854.2425
25      2020-08  product_E      0.0000

In [12]:
contadorA = 0
contadorB = 0
contadorC = 0
contadorD = 0
contadorE = 0

for i in range(len(df_merge['product_A'])):
    contadorA += df_merge['product_A'][i]

for i in range(len(df_merge['product_B'])):
    contadorB += df_merge['product_B'][i]

for i in range(len(df_merge['product_C'])):
    contadorC += df_merge['product_C'][i]

for i in range(len(df_merge['product_D'])):
    contadorD += df_merge['product_D'][i]

for i in range(len(df_merge['product_E'])):
    contadorE += df_merge['product_E'][i]


contadorA, contadorB, contadorC, contadorD, contadorE

(58512, 68856, 142776, 102936, 15000)

In [13]:
contA = 0
contB = 0
contC = 0
contD = 0
contE = 0

for i in range(len(df_mark['product'])):
    if(df_mark['product'][i] == 'product_A'):
        contA += df_mark['investment'][i]
    if(df_mark['product'][i] == 'product_B'):
        contB += df_mark['investment'][i]
    if(df_mark['product'][i] == 'product_C'):
        contC += df_mark['investment'][i]
    if(df_mark['product'][i] == 'product_D'):
        contD += df_mark['investment'][i]
    if(df_mark['product'][i] == 'product_E'):
        contE += df_mark['investment'][i]

contA, contB, contC, contD, contE

(172594.75, 92285.64000000001, 265173.6325, 84385.0625, 0.0)

### 2. Observando como as preocupações e seus pesos relativos se movimentam ao longo do período analisado na resposta do item 1:
- #### Quais são as hipóteses que poderiam justificar esses movimentos?

Pode-se observar nas variações mês a mês que existe uma sazionalidade das preocupações, onde as preocupações com a pele aumentam em determinados períodos do ano, sendo o principal período durante os meses 4 ao mês 9. A hipótese para este movimento de sazionalidade pode ser explicada pela temperatura e pelo clima predominante nesta estação do ano, pois o tempo seco e as temperaturas baixas podem aumentar as preocupações existentes com a pele, como o ressecamento da pele e consequentemente trazer outras preocupações. Desta forma, a temperatura e o clima da estação podem influenciar nas preocupações da pele.

De tal modo, em 2020 também tivemos da pandemia do COVID-19, onde muitas pessoas tiveram que ficar em casa. Pode-se observar assim, que no começo da pandemia existiu um grande aumento das preocupações com a pele, podendo ser influenciado por causa que as pessoas precisaram ficar em casa isoladas e, analisando esse fato, pode-se dizer que tiveram mais tempo para cuidar da saúde de suas peles. Assim, com a análise mais cuidadosa de suas peles, preocupações recorrentes ou não podem aparecer e assim influenciar em suas submissões. Desta forma, pode-se concluir que durante a pandemia existiu mais um movimento de preocupação maior com a pele.

Também analisando o investimento na campanha de marketing em cada produto, pode-se perceber que produtos que tiveram maior investimento obtiveram maior quantidade de compras, desta forma, é verídico qu existe um movimento relacionado à quantidade de investimento feita em determinados produtos, desta forma, quanto mais investimentos em determinados produtos, mais compras podem existir deste produto. A tabela abaixo exemplificará o movimento que quero dizer:

Produto  | Investimento    | Compras    
-------- | --------------- | -----------
A        | 172594.75       | 58512  
B        | 92285.64        | 68856  
C        | 265173.63       | 142776  
D        | 84385.06        | 102936  
E        | 0.0             | 15000  

Concluo assim, que o movimento de compra a partir dos investimentos também é uma hipótese de movimento que deve ser analisada mais a fundo, pois com ela existe a possibilidade de melhoria dos rendimentos da empresa.

- #### Como você validaria essas hipóteses?

Algumas hipóteses seriam mais difíceis de validar devido a complexidade de seus dados, como os movimentos da pandemia, por exemplo, são dados que já foram e teria que ser realizadas pesquisas para validá-los e provavelmente não seria algo que traria alguma clareza ou, talvez, seria eficiente caso alguma pesquisa de consumo aborde esse tema de ter um tempo para cuidar mais da pele, então talvez seria interessante validar para comprovar esse movimento para poder tirar algum lucro sobre essa hipótese.

A hipótese da questão de gerar mais compras a partir dos investimentos em publicidade pode ser validada como fiz acima, visualizando os valores e analisando se o retorno com compras existiu, desta forma validando que os investimentos em campanhas de marketing gera um retorno em compras e cria uma influência das respostas do quiz. Diretamente ou indiretamente, o desejo de compra gera uma preocupação com a falta daquilo que queremos comprar, ou seja, se a pessoa quer um produto, ela provavelmente vai sentir que seu corpo precisa, o psicológico influencia nas compras. Desta forma, conclui-se, que boas publicidades podem gerar boas compras e preocupações a mais com a pele.

Movimentos relacionados a sazionalidades podem ser validados por meio de análises dos dados da empresa, juntamente com outros dados de outras pesquisas relacionadas à qualidade e situação da pele durante os meses do ano, podendo prever, assim, o que pode ser mais vendido em determinadas épocas do ano e o que pode ser feito para melhorar as vendas de outros produtos nos demais meses do ano. Nessas outras pesquisas, os dados podem ser coletados por formulários ou por parcerias com dermatologistas que verifiquem a qualidade da pele e o interesse dos pacientes em cuidar da pele, podendo assim ter uma coleta de dados mais segura. Desta forma pode-se comprovar essa hipótese.

Demais movimentos que podem vir a existir no ambiente do dados, como questões financeiras que podem variar conforme época do ano ou até mesmo em determinados anos, que venham a reduzir ou ampliar as preocupações e compras, podem ser validados por ampliação da coleta dos dados do quiz, colocando alguns termos a mais para tentar identificar e validar o que pode fazer com que as demais variações ocorram.